### PR TITLE
3.1.0 release

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

**Comment by guolinke** [comment]
I think following PRs could be included in this release:

#2754
#3425 
#3428
#3485

for the remaining, we may need more times to review/implement.

**Comment by StrikerRUS** [comment]
I think #3428 should be merged only after fixing this CUDA error #3450.

Also, I'd like to include in `3.1` a fix for easy-reproducible OpenCL GPU error #3475 .

**Comment by jameslamb** [comment]
Works for me! Other than the pull requests already mentioned, I have nothing to add that I think needs to get in for the release. And I agree with the next release being 3.1.0 instead of 3.0.1, since so many new features have been added.

Once those other pull requests mentioned above have all been merged, I'll run all the CRAN checks (manual `valgrind` check here, Solaris via R Hub) to be sure no new issues have been introduced.

We won't be able to submit the R package to CRAN until November but I'm ok with doing the rest of the release before then and not waiting on CRAN. Also as a reminder from #3338 ...I'll do the CRAN submission this time.

**Comment by jameslamb** [comment]
@guolinke I hope it's ok, I updated `R-package/configure` and pushed it here

**Comment by StrikerRUS** [comment]
### Release checklist:

- [ ] Update [VERSION](https://github.com/microsoft/LightGBM/blob/master/VERSION.txt) number.
- [ ] Update version in [Appveyor config file](https://github.com/microsoft/LightGBM/blob/master/.appveyor.yml).
- [ ] Update version and release date in [`DESCRIPTION` file of R-package](https://github.com/microsoft/LightGBM/blob/master/R-package/DESCRIPTION).
- [ ] Push commit with `v*` tag to trigger `GitHubRelease` action at Azure Pipelines.
- [ ] Convert automatic GitHub release from `Draft` to normal one.
- [ ] Update [`stable` tag at GitHub](https://github.com/microsoft/LightGBM/releases/tag/stable).
- [ ] Upload release to [PyPI](https://pypi.org/project/lightgbm/).
- [ ] Upload release to CRAN (soon).
- [ ] Upload release to [NuGet](https://www.nuget.org/packages/LightGBM/).
- [ ] Update version and hash in [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/lightgbm.rb).

**Comment by jameslamb** [comment]
As of #3335, `Update version and release date in DESCRIPTION file of R-package.` is no longer necessary for releases 😀

**Comment by StrikerRUS** [comment]
@jameslamb Cool! Then I'm striking this bullet out from the list.

**Comment by guolinke** [comment]
@shiyu1994 what is the ETA of your improvement for dense histogram construction? I think this PR is waiting for it.

**Comment by StrikerRUS** [comment]
I've marked #3520 as blocking issue, I think we can include fix for it in `3.1.0`.

**Comment by jameslamb** [comment]
I'm sorry, I just discovered a valgrind issue that I think will block 3.1.0 from getting to CRAN: #3525

**Comment by jameslamb** [comment]
Now that all of the PRs in https://github.com/microsoft/LightGBM/pull/3484#issuecomment-716306645 have been fixed,  if CI passes are we ready to release? Should I submit the R package to CRAN?

File: R-package/configure
Comment (jameslamb) [inline R-package/configure:3]: /gha run r-valgrind

**Comment by jameslamb** [comment]
oh wait I forgot about #3513. I'll open a PR right now that skips the problematic tests for now.

File: R-package/configure
Comment (jameslamb) [reply R-package/configure:3]: this passed: https://github.com/microsoft/LightGBM/actions/runs/350733081

**Comment by StrikerRUS** [comment]
> if CI passes are we ready to release?

What about https://github.com/microsoft/LightGBM/pull/3484#issuecomment-716507553 (#3475)?

**Comment by StrikerRUS** [comment]
Also, I'd like to mark this release as the last one which supports Python 2. So that our next minor release (in 2021 I believe) will drop Python 2 support (1 year after Python 2 EOL). I guess it's time. Does anyone have any objections?

**Comment by jameslamb** [comment]
> > if CI passes are we ready to release?
> 
> What about [#3484 (comment)](https://github.com/microsoft/LightGBM/pull/3484#issuecomment-716507553) (#3475)?

oh sorry, missed that one. Yes we should try to fix that before release.

> Also, I'd like to mark this release as the last one which supports Python 2. So that our next minor release (in 2021 I believe) will drop Python 2 support (1 year after Python 2 EOL). 

I agree with this too. I think it's ok to abandon support for Python 2. Other scientific computing libraries are already talking about even removing support for Python 3.6:

* https://github.com/dask/community/issues/66
* https://numpy.org/neps/nep-0029-deprecation_policy.html

**Comment by shiyu1994** [comment]
> @shiyu1994 what is the ETA of your improvement for dense histogram construction? I think this PR is waiting for it.

Done with that.

**Comment by guolinke** [comment]
I think it is almost ready. Any other PRs need to include?

**Comment by jameslamb** [comment]
@StrikerRUS recommended that we try to fix #3475 before release

**Comment by StrikerRUS** [comment]
One more good item to include in new release, I believe:

- [ ] #3561

**Comment by guolinke** [comment]
> @StrikerRUS recommended that we try to fix #3475 before release

I saw @StrikerRUS already try different python and boost versions, and think it is not trivial to fix in a short time. 
And it is possible that the bug is from Boost or Clang, as the gcc and msvc can work.

File: .appveyor.yml
Comment (guolinke) [inline .appveyor.yml:1]: /gha run r-valgrind

Comment (StrikerRUS) [reply .appveyor.yml:1]: valgrind is OK.
https://github.com/microsoft/LightGBM/runs/1402192963?check_suite_focus=true

@jameslamb 
Seems we should update the number of bytes in out script

> 336 bytes in 1 blocks are possibly lost

https://github.com/microsoft/LightGBM/blob/742d72f8bb051105484fd5cca11620493ffb0b2b/.ci/test_r_package_valgrind.sh#L63

**Review by StrikerRUS — APPROVED** [review]

**Comment by guolinke** [comment]
I am going to merge this in 24 hours and draft a new release.

**Comment by guolinke** [comment]
@jameslamb Any cran's issuse to address? I think we can submit this release to cran.

File: .appveyor.yml
Comment (jameslamb) [reply .appveyor.yml:1]: 👍 ok I can do that. It doesn't have to be done before this release though.

**Comment by jameslamb** [comment]
> @jameslamb Any cran's issuse to address? I think we can submit this release to cran.

I'll try R Hub Solaris checks on this branch one more time right now. I think some things have been merged here since the last time I did that.

If those checks pass (should only take 10-20 minutes), I'll submit a CRAN submission. Is that OK?

File: R-package/configure
Comment (jameslamb) [reply R-package/configure:3]: /gha build r-artifacts

Comment (jameslamb) [reply R-package/configure:3]: looks like this job is broken:

> Error: Input 'submodules' not supported when falling back to download using the GitHub REST API. To create a local Git repository instead, add Git 2.18 or higher to the PATH.

run: https://github.com/microsoft/LightGBM/runs/1404264739?check_suite_focus=true

I opened https://github.com/microsoft/LightGBM/issues/3567. It doesn't have to be fixed for the release. I can build the package locally.

**Comment by guolinke** [comment]
@jameslamb no problem, thank you!

**Comment by jameslamb** [comment]
no issues found on Solaris (https://artifacts.r-hub.io/lightgbm_3.1.0.tar.gz-048076e707fd42c695b805753f65aa44/lightgbm.Rcheck/), `valgrind` is happy (https://github.com/microsoft/LightGBM/pull/3484#discussion_r523766573), and all of the other CI jobs are passing.

I'm going to submit to CRAN 😀 

@guolinke I think that since I'm not the maintainer in `DESCRIPTION`, CRAN will email you to confirm that I have permission to submit.

**Comment by jameslamb** [comment]
I've submitted to CRAN!

image

@guolinke they'll send you an email next asking you to confirm the submission

> The maintainer of this package has been sent an email to confirm the submission. After their confirmation the package will be passed to CRAN for review.

> Please note that the submission is not completed until it is confirmed by the maintainer (via a link in the e-mail), who may need to check the spam folder.

**Comment by guolinke** [comment]
> ### Release checklist:
> * [x]  Update [VERSION](https://github.com/microsoft/LightGBM/blob/master/VERSION.txt) number.
> * [x]  Update version in [Appveyor config file](https://github.com/microsoft/LightGBM/blob/master/.appveyor.yml).
> * [x]  ~Update version and release date in [`DESCRIPTION` file of R-package](https://github.com/microsoft/LightGBM/blob/master/R-package/DESCRIPTION).~
> * [x]  Update version in [`configure` file of R-package](https://github.com/microsoft/LightGBM/blob/master/R-package/configure).
> * [x]  Push commit with `v*` tag to trigger `GitHubRelease` action at Azure Pipelines.
> * [x]  Convert automatic GitHub release from `Draft` to normal one.
> * [x]  Update [`stable` tag at GitHub](https://github.com/microsoft/LightGBM/releases/tag/stable).
> * [x]  Upload release to [PyPI](https://pypi.org/project/lightgbm/).
> * [ ]  Upload release to CRAN (soon).
> * [x]  Upload release to [NuGet](https://www.nuget.org/packages/LightGBM/).
> * [ ]  Update version and hash in [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/lightgbm.rb).

@StrikerRUS can you help for the  Homebrew formula?

**Comment by StrikerRUS** [comment]
@guolinke 
>can you help for the Homebrew formula?

Sure. PR proposed: https://github.com/Homebrew/homebrew-core/pull/64933

Seems Homebrew started to support Big Sur! 👍 

https://github.com/Homebrew/homebrew-core/commit/d80a8b51f83393dae4c6222077363a77dc78435e#diff-5e461287a37b620a7dd2f18a9c1706d10371608f3384c032bb732aa2a08117f6

**Comment by StrikerRUS** [comment]
> PR proposed: Homebrew/homebrew-core#64933

Homebrew successfully updated:
- https://github.com/Homebrew/homebrew-core/commit/133a0441bef47e9d4f9f91980ccde7512c604e21#diff-5e461287a37b620a7dd2f18a9c1706d10371608f3384c032bb732aa2a08117f6
- https://github.com/Homebrew/homebrew-core/commit/6c9e77fc0e3b3ff51b26a89c80dbf385f4bba230#diff-5e461287a37b620a7dd2f18a9c1706d10371608f3384c032bb732aa2a08117f6

**Comment by jameslamb** [comment]
woo, thanks!

**Comment by jameslamb** [comment]
@guolinke have you received an email from CRAN?

According to https://lockedata.github.io/cransays/articles/dashboard.html, `{lightgbm}` is still in "pending" which means CRAN is waiting for a response from us. They should have sent you an email with a link to click to confirm that I'm allowed to submit.

image

**Comment by guolinke** [comment]
I think I already comfirm it and receive their auto reply:
```
Dear maintainer,
 
package lightgbm_3.1.0.tar.gz has been auto-processed and is pending a manual inspection of this new CRAN submission. A CRAN team member will typically respond to you within the next 10 working days. For technical reasons you may receive a second copy of this message when a team member triggers a new check.
 
Log dir: <https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwin-builder.r-project.org%2Fincoming_pretest%2Flightgbm_3.1.0_20201116_044056%2F&amp;data=04%7C01%7Cguolin.ke%40microsoft.com%7Cbe31c3b4f887472079fe08d88a006d6d%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637411081749888900%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&amp;sdata=NZe%2BwpgRf5A%2FTfQy1OZIIvl2y1DORnbdUFb4YZyZ5N8%3D&amp;reserved=0>
The files will be removed after roughly 7 days.
Installation time in seconds: 676
Check time in seconds: 256 
R Under development (unstable) (2020-11-14 r79432)
 
Pretests results:
Windows: <https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwin-builder.r-project.org%2Fincoming_pretest%2Flightgbm_3.1.0_20201116_044056%2FWindows%2F00check.log&amp;data=04%7C01%7Cguolin.ke%40microsoft.com%7Cbe31c3b4f887472079fe08d88a006d6d%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637411081749888900%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&amp;sdata=MfQWw19iM7fjFNw5HhQXZtB4LmyLvHRMuiIQeaNWl%2B8%3D&amp;reserved=0>
Status: 1 NOTE
Debian: <https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwin-builder.r-project.org%2Fincoming_pretest%2Flightgbm_3.1.0_20201116_044056%2FDebian%2F00check.log&amp;data=04%7C01%7Cguolin.ke%40microsoft.com%7Cbe31c3b4f887472079fe08d88a006d6d%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637411081749898895%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&amp;sdata=nYwviy5FXgrR0wgBAeFd1aGe%2BC8f3WnK%2BAz4ZhBRV5A%3D&amp;reserved=0>
Status: 1 NOTE
 

 
No strong reverse dependencies to be checked.
 
Best regards,
CRAN teams' auto-check service

```

**Comment by jameslamb** [comment]
oh great, thank you!

**Comment by guolinke** [comment]
It seems we got accept: 
https://cran.r-project.org/web/packages/lightgbm/index.html

**Comment by jameslamb** [comment]
!!! Exciting!

I'm going to wait a few days before publicizing this or updating the documentation, in case we get surprised like last time.

But I'm hopeful 😀

**Comment by jameslamb** [comment]
I've been refreshing https://cran.r-project.org/web/checks/check_results_lightgbm.html every few hours. More checks are slowly rolling in.

Right now we're passing on Windows, Mac, Solaris, and Linux

image

Soon we should know if we're passing `valgrind` and `UBSAN/ASAN`, as well as tests with other combinations of operating system + compiler.

**Comment by StrikerRUS** [comment]
@jameslamb According to https://cran.r-project.org/web/checks/check_results_lightgbm.html it looks like we need to add `ON_32BIT_WIN` to the fix introduced in #3534.

**Comment by jameslamb** [comment]
UGH



Ok I'll open a PR with a fix. I think to be safe we should just completely skip that test if on CRAN. This also tells me the existing Windows CI jobs we have in GitHub Actions aren't a perfect replica of what CRAN does.

I'll see if I can replicate it with R Hub.

**Comment by jameslamb** [comment]
I created #3585 , let's move this discussion there instead of keeping it on this closed PR

**Comment by StrikerRUS** [comment]
> This also tells me the existing Windows CI jobs we have in GitHub Actions aren't a perfect replica of what CRAN does.

I guess we simply don't have any 32-bit CI jobs.

**Pull request merged**
